scripts/o3de/o3de/repo_properties.py

```diff
@@ -0,0 +1,377 @@
+    :param release_archive_path: Optional local path to a folder where a release archive should be written
+    :param download_prefix: The prefix of the download uri
+    """
+    # The beginning remote repo json template
+    repo_objects_data = repo_json.get(f'{object_typename}s_data',[])
+   
+    if add_objects or replace_objects:
+        if replace_objects:
+            repo_objects_data = []
+            paths = replace_objects.split() if isinstance(replace_objects, str) else replace_objects
+        else:
+            paths = add_objects.split() if isinstance(add_objects, str) else add_objects
+        for object_path in paths:
+            # object JSON data of the object you want to add
+            json_data = manifest.get_json_data(object_typename, object_path, validator)
+            if json_data:
+                if release_archive_path:
+                    version = json_data.get('version','0.0.0')
+                    archive_filename = pathlib.PurePath(f"{json_data[f'{object_typename}_name']}-{version}-{object_typename}.zip".lower())
+                    json_data = create_remote_object_archive(object_path, 
+                                object_path / f'{object_typename}.json', 
+                                archive_filename, 
+                                release_archive_path, 
+                                repo_json['repo_uri'], 
+                                download_prefix)
+                index = _find_index(repo_objects_data, f'{object_typename}_name', json_data[f'{object_typename}_name'])
+                # index will be non-negative if an instance of this object already exists in the repo.json
+                if index != -1:
+                    # we want changes only
+                    version_data = _changed(repo_objects_data[index], json_data)
+                    if not version_data:
+                        # there is no difference between the object being added and 
+                        # what already exists in repo.json
+                        logger.warning(f"No changes were found between the object JSON data in repo.json and the object data being added from {object_path}")
+                        continue
+
+                    # versions data must always include the version
+                    version_data['version'] = json_data.get('version','0.0.0')
+
+                    # versions data must always include one of the download/source uris
+                    if 'download_source_uri' in json_data:
+                        version_data['download_source_uri'] = json_data['download_source_uri']
+                    if 'source_control_uri' in json_data:
+                        version_data['source_control_uri'] = json_data['source_control_uri']
+                    if 'source_control_ref' in json_data:
+                        version_data['source_control_ref'] = json_data['source_control_ref']
```

Comment:
Should there be a check for if none of the uris are provided?

Reply:
We should add a warning here in case there isn't at least a `download_source_uri` or `source_control_uri`.  In theory there could be a `download_source_uri` in the gem data and not in the `versions_data` list that it could fall back to, but I think that would almost certainly not be what the user intends.

Reply:
source_control_uri isn't being used in o3de-extras. is there a working of this? assuming it's just the github.com path to the object's (gem/template/project) folder? or maybe it's a https://raw.githubusercontent.com/ link?

Reply:
it's not required, you just have to have `download_source_uri` or `source_control_uri` or both.  For o3de-extras we prefer to use a downloadable archive because if we used the source control method we would be cloning multiple copies of the entire repo.